File: Solutions/1530- Number of Good Leaf Nodes Pairs/Solution.py
class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        global count
        count = 0
        self.dfs(root,distance)
        return count

    def dfs(self, root, distance):
        global count
        if root is None:
            return []
        if root.left is None and root.right is None:
            return [1]

        left = self.dfs(root.left, distance)
        right = self.dfs(root.right, distance)

        for l in left:
            for r in right:
                if l + r <= distance:
                    count += 1

        leaves = list()
        for l in left:
            if l + 1 < distance:
                leaves.append(l + 1)

        for r in right:
            if r + 1 < distance:
                leaves.append(r + 1)

        return leaves

File: Solutions/1530- Number of Good Leaf Nodes Pairs/Solution.java
class Solution {
    int count = 0;

    public int countPairs(TreeNode root, int distance) {
        dfs(root,distance);
        return count;
    }

    private List<Integer> dfs(TreeNode root, int distance) {
        if (root == null)
            return new ArrayList<>();

        if (root.left == null && root.right == null) {
            List<Integer> temp = new ArrayList<>();
            temp.add(1);
            return temp;
        }

        List<Integer> left = dfs(root.left, distance);
        List<Integer> right = dfs(root.right, distance);

        for (int l : left) {
            for (int r : right) {
                if (l + r <= distance)
                    count++;
            }
        }

        List<Integer> leaves = new ArrayList<>();
        
        for (int l : left) {
            if (l + 1 < distance)
                leaves.add(l + 1);
        }
        
        for (int r : right) {
            if (r + 1 < distance)
                leaves.add(r + 1);
        }

        return leaves;
    }
}